### PR TITLE
[115] Access Table elements in table expressions and actions

### PR DESCRIPTION
**Pull request by mypsycho**
Ticket #115

### CODE REVIEW

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/print/PrintAsImageCallback.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2010, 2015 THALES GLOBAL SERVICES.
+ * Copyright (c) 2010, 2015, 2023 THALES GLOBAL SERVICES.
```
Comment (lredor) [inline plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/print/PrintAsImageCallback.java:2]: 2010, 2023

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DTableColumnLabelProvider.java [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2007, 2016 THALES GLOBAL SERVICES.
+ * Copyright (c) 2007, 2016, 2023 THALES GLOBAL SERVICES.
```
Comment (lredor) [inline plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DTableColumnLabelProvider.java:2]: 2007, 2023

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DTableColumnLabelProvider.java [hunk 3]
```diff
@@ -225,6 +225,9 @@ protected Option<DCell> getDCell(final Object element) {
 
     /**
      * Returns the image with the given key, or <code>null</code> if not found.
+     * 
+     * @param isSelected
```
Comment (lredor) [inline plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DTableColumnLabelProvider.java:229]: It's only a "happy checkstyle" here, without interest. The parameter must be documented.

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/DTableViewerManager.java
Comment (lredor) [inline plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/DTableViewerManager.java:1]: Maybe the commit message should be completed: "[cleanup] Use new Event API for SWT key or mouse listeners."

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java [hunk 4]
```diff
@@ -601,8 +601,6 @@ public void refreshStyle(final DCell cell) {
      * @param style
```
Comment (lredor) [inline plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java:601]: For comment of the commit,  explanations are welcome. In what do you simplify this class? The "simplification" only concerns "refactoring" or is there something change on the logic ?

The comment title could be a little bit more explicit: "[cleanup] Simplify style update of DTableElementSynchronizer"

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java [hunk 5]
```diff
@@ -781,10 +797,10 @@ private void doUpdateStyle(final DColumn column, final DTableElementStyle style)
         if (bestForegroundStyle != null) {
             boolean defaultStyleDescription = new StyleUpdaterQuery(styleUpdater).isDefaultForegroundColor(bestForegroundStyle.getForeGroundColor());
             colorUpdater.updateForegroundColor(style, bestForegroundStyle.getForeGroundColor(), defaultStyleDescription, column.getTarget());
-            if (bestForegroundStyle.getLabelFormat() != null && !Objects.equals(style.getLabelFormat(), bestForegroundStyle.getLabelFormat())) {
+            if (!Objects.equals(style.getLabelFormat(), bestForegroundStyle.getLabelFormat())) {
                 FontFormatHelper.setFontFormat(style.getLabelFormat(), bestForegroundStyle.getLabelFormat());
             }
-            if (bestForegroundStyle.getLabelSize() != -1 && style.getLabelSize() != bestForegroundStyle.getLabelSize()) {
+            if (style.getLabelSize() != bestForegroundStyle.getLabelSize()) {
```
Comment (lredor) [inline plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java:803]: You remove "bestForegroundStyle.getLabelSize() != -1" but there is maybe a link with auto size. Have you checked ? Furthermore, you not remove it for the doUpdateStyle of a DLine.

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java [hunk 6]
```diff
@@ -17,22 +17,25 @@
 import java.util.Collections;
```
Comment (lredor) [inline plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java:17]: There are a lot of modifications in this class. It's difficult to review without missing anything. I think that a code coverage must be done before and after changes with EclEmma using suite:
org.eclipse.sirius.tests.suite.table.AllTablePluginTests
org.eclipse.sirius.tests.suite.table.AllTableStandaloneTests
and org.eclipse.sirius.tests.swtbot.suite.TableSwtbotTestSuite
If the coverage is not of 100%, I think that additional tests must be added in the context of these changes.

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableSynchronizerImpl.java [hunk 7]
```diff
@@ -17,25 +17,24 @@
 import java.util.HashMap;
```
Comment (lredor) [inline plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableSynchronizerImpl.java:17]: Same remarks than for DTableElementSynchronizer

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DFeatureColumnEditingSupport.java [hunk 8]
```diff
@@ -153,15 +152,30 @@ protected boolean canEdit(final Object element) {
...
-                    } catch (final EvaluationException e) {
-                        RuntimeLoggerManager.INSTANCE.error(updater, DescriptionPackage.eINSTANCE.getCellUpdater_CanEdit(), e);
-                    }
+                if (updater != null && updater.getCanEdit() != null && !updater.getCanEdit().isEmpty()) {
+
+                    canEdit = InterpretationContext.with(cell.getTarget(), ctx -> {
+                        DTable table = TableHelper.getTable(line);
+                        
+                        ctx.setVariable(IInterpreterSiriusVariables.CONTAINER_VIEW, line);
+                        ctx.setVariable(IInterpreterSiriusTableVariables.LINE, line);
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DFeatureColumnEditingSupport.java:164]: It mimics the existing statement that follows where 'container' and 'lineSemantic' target the same value "line.getTarget()".

[the remark this answers is not in the document]
Comment (lredor) [reply plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DFeatureColumnEditingSupport.java:164]: Indeed.

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java [hunk 9]
```diff
@@ -387,17 +387,94 @@ private boolean refreshLabelIntersectionMapping(final DCell cell, final String l
...
+     *            Semantic element to find from
+     * @param container
+     *            the table or line to search from (accordingly with using domain class)
+     * @param iMapping
+     *            The intersection mapping that contains the expression to evaluate
+     * @param logError
+     *            Flag to log if an error happens
+     * @return List of column candidates for this cell and this intersection mapping.
+     */
+    public Collection<EObject> evaluateColumnFinderExpression(EObject candidate, DSemanticDecorator container, IntersectionMapping iMapping, boolean logError) {
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java]: Done

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java [hunk 10]
```diff
@@ -387,17 +387,94 @@ private boolean refreshLabelIntersectionMapping(final DCell cell, final String l
      * @return List of column candidates for this cell and this intersection mapping.
      */
     private Collection<EObject> evaluateColumnFinderExpression(DCell cell, IntersectionMapping iMapping) {
-        EObject preconditionContext;
+        DSemanticDecorator source;
+        DSemanticDecorator container;
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java:368]: Done

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java [hunk 11]
```diff
@@ -387,17 +387,94 @@ private boolean refreshLabelIntersectionMapping(final DCell cell, final String l
...
         if (iMapping.isUseDomainClass()) {
-            preconditionContext = cell.getTarget();
+            source = cell;
+            container = TableHelper.getTable(cell.getLine());
         } else {
-            preconditionContext = cell.getLine().getTarget();
+            source = cell.getLine();
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java:373]: Done

File: plugins/org.eclipse.sirius/src/org/eclipse/sirius/business/internal/logger/RuntimeLoggerInterpreterImpl.java [hunk 12]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2009, 2021 THALES GLOBAL SERVICES.
+ * Copyright (c) 2009, 2021, 2023 THALES GLOBAL SERVICES.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius/src/org/eclipse/sirius/business/internal/logger/RuntimeLoggerInterpreterImpl.java:2]: Done

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/tests/srs/ExploitArrays_SimpleTest.java [hunk 13]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2010, 2019 THALES GLOBAL SERVICES.
+ * Copyright (c) 2010, 2019, 2023 THALES GLOBAL SERVICES.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/tests/srs/ExploitArrays_SimpleTest.java:2]: Done

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/tests/srs/REQ_157_Test.java [hunk 14]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2010, 2014 THALES GLOBAL SERVICES.
+ * Copyright (c) 2010, 2014, 2023 THALES GLOBAL SERVICES.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/tests/srs/REQ_157_Test.java:2]: Done

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/tests/srs/REQ_161_Test.java [hunk 15]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2010, 2014 THALES GLOBAL SERVICES.
+ * Copyright (c) 2010, 2014, 2023 THALES GLOBAL SERVICES.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/tests/srs/REQ_161_Test.java:2]: Done

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/suite/table/AllTablePluginTests.java [hunk 16]
```diff
@@ -109,6 +109,9 @@ public static Test suite() {
         suite.addTestSuite(TableStyleRefreshTest.class);
         suite.addTestSuite(TableWithMultivaluedAttributeTest.class);
         suite.addTestSuite(TableWithHeaderColumnWidthTest.class);
+        suite.addTestSuite(SelectionInTableAfterToolExecutionTest.class);
+        // FIXME this test seems out-dated because of AIRD evolution.
+        // suite.addTestSuite(TableNotDirtyOnOpeningTest.class);
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/suite/table/AllTablePluginTests.java:114]: Fixed: the refresh would create invisible empty cell. The test is now restored.

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/tests/srs/SpecifyArrays_Test.java [hunk 17]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2010, 2014 THALES GLOBAL SERVICES.
+ * Copyright (c) 2010, 2014, 2023 THALES GLOBAL SERVICES.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/tests/srs/SpecifyArrays_Test.java:2]: Done

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/common/TableTestCase.java [hunk 18]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2010, 2014 THALES GLOBAL SERVICES.
+ * Copyright (c) 2010, 2014, 2023 THALES GLOBAL SERVICES.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/common/TableTestCase.java:2]: Done

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/refresh/DTableCellStylesTest.java [hunk 19]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2010, 2015 THALES GLOBAL SERVICES and others.
+ * Copyright (c) 2010, 2015, 2023 THALES GLOBAL SERVICES and others.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/refresh/DTableCellStylesTest.java:2]: Done

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/refresh/DTableSynchronizerTest.java [hunk 20]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2010, 2015 THALES GLOBAL SERVICES and others.
+ * Copyright (c) 2010, 2015, 2023 THALES GLOBAL SERVICES and others.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/refresh/DTableSynchronizerTest.java:2]: Done

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/sort/DTableSortByLineTest.java [hunk 21]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2010, 2014 THALES GLOBAL SERVICES.
+ * Copyright (c) 2010, 2014, 2023 THALES GLOBAL SERVICES.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/sort/DTableSortByLineTest.java:2]: Done

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/suite/table/AllTablePluginTests.java [hunk 22]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2010, 2018 THALES GLOBAL SERVICES and others.
+ * Copyright (c) 2010, 2018, 2023 THALES GLOBAL SERVICES and others.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/suite/table/AllTablePluginTests.java:2]: Done

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DFeatureColumnEditingSupport.java [hunk 23]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2008, 2021 THALES GLOBAL SERVICES and others.
+ * Copyright (c) 2008, 2021, 2023 THALES GLOBAL SERVICES and others.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DFeatureColumnEditingSupport.java:2]: Done

File: plugins/org.eclipse.sirius/src/org/eclipse/sirius/business/api/logger/InterpretationContext.java [hunk 24]
```diff
@@ -0,0 +1,173 @@
+/*******************************************************************************
+ * Copyright (c) 2022, 2023 THALES GLOBAL SERVICES and others.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius/src/org/eclipse/sirius/business/api/logger/InterpretationContext.java:2]: Done

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DTargetColumnEditingSupport.java [hunk 25]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2007, 2008 THALES GLOBAL SERVICES.
+ * Copyright (c) 2007, 2008, 2023 THALES GLOBAL SERVICES.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DTargetColumnEditingSupport.java:2]: Done

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/api/helper/TableHelper.java [hunk 26]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2007, 2021 THALES GLOBAL SERVICES and others.
+ * Copyright (c) 2007, 2021, 2023 THALES GLOBAL SERVICES and others.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/api/helper/TableHelper.java:2]: Done

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/commands/SortDLinesCommand.java [hunk 27]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2012, 2015 THALES GLOBAL SERVICES.
+ * Copyright (c) 2012, 2015, 2023 THALES GLOBAL SERVICES.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/commands/SortDLinesCommand.java:2]: Done

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/DTableEraseItemListener.java [hunk 28]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2009, 2009 THALES GLOBAL SERVICES.
+ * Copyright (c) 2009, 2009, 2023 THALES GLOBAL SERVICES.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/DTableEraseItemListener.java:2]: Done

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/action/ShowAllLinesAction.java [hunk 29]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2008, 2015 THALES GLOBAL SERVICES.
+ * Copyright (c) 2008, 2015, 2023 THALES GLOBAL SERVICES.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/action/ShowAllLinesAction.java:2]: Done

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/sort/DTableSortByColumnTest.java [hunk 30]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2010, 2014 THALES GLOBAL SERVICES.
+ * Copyright (c) 2010, 2014, 2023 THALES GLOBAL SERVICES.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/sort/DTableSortByColumnTest.java:2]: Done

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/listeners/DLineExpansionChecker.java [hunk 31]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2011, 2020 THALES GLOBAL SERVICES.
+ * Copyright (c) 2011, 2020, 2023 THALES GLOBAL SERVICES.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/listeners/DLineExpansionChecker.java:2]: Done

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DTableColumnLabelProvider.java (hunk 2 above)
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DTableColumnLabelProvider.java:2]: Done``

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/print/PrintAsImageCallback.java (hunk 1 above)
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/print/PrintAsImageCallback.java:2]: Done

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DTableColumnLabelProvider.java (hunk 3 above)
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/provider/DTableColumnLabelProvider.java:229]: Done

File: plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/DTableViewerManager.java
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table.ui/src/org/eclipse/sirius/table/ui/tools/internal/editor/DTableViewerManager.java:1]: Done

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java (hunk 5 above)
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java:803]: Handling labelSize  == -1 is performed by UI : org.eclipse.sirius.table.ui.tools.internal.editor.provider.DTableColumnLabelProvider
This is now also applied to doUpdateStyle(DLine, ...).

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java (hunk 4 above)
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java:601]: It is mistake: this commit should belong to ticket #116.

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java (hunk 6 above)
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java:17]: Code coverage gives the same level of coverage before and after the refactoring.
Following segments of case were/are not covered:

In CrossTableSynchronizer
- if no interactions, remove any existing cells.
- refreshColumnCandidate when position changes in description, it is reflect in the table.
- remove columns without target.
- Evaluate semantic candidates of IntersectionMapping without domain.
- updateCells: handle mapping change and/or target change.

In AbstractTableSynchronizer
- detectLineToRefresh : moving order of lines.
- remove line without targets.

In EditionTableSynchronizer
- remove cell without lines.
- removeOldCellsInTable : this is probably dead code. there is a comment about it.
- refresh : reach cellStillExists == false.
- refreshColumnMapping : moving columns : adding/removing column.

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableSynchronizerImpl.java (hunk 7 above)
Comment (mypsycho) [reply plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableSynchronizerImpl.java:17]: Code coverage gives the same level of coverage before and after the refactoring.
Following segments of case were/are not covered:

In DTableElementSynchronizer
- column with default width == 0
- line without target
- refreshTarget for Cell of CrossTable

**Comment by mypsycho** [comment]
There is a sample in org.eclipse.sirius.tests.junit using 'table' variable in different places.
/org.eclipse.sirius.tests.junit/data/table/unit/variables/tables.uml is the model.
/org.eclipse.sirius.tests.junit/data/table/unit/variables/TableVariable.odesign.

In the table description, we use annotation entries to create lines targets.
We need 'table' variables to create such virtual lines: lines that are not based on elements of the model but on result of some services.

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java (hunk 6 above)
Comment (lredor) [reply plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableElementSynchronizer.java:17]: The list of tests to add could be done later. But it's an intersting list to increase the code coverage.

File: plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableSynchronizerImpl.java (hunk 7 above)
Comment (lredor) [reply plugins/org.eclipse.sirius.table/src/org/eclipse/sirius/table/business/internal/refresh/DTableSynchronizerImpl.java]: Same remark (to be done later)

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/variables/VirtualTableService.java [hunk 32]
```diff
@@ -0,0 +1,173 @@
...
+        result.getDetails().put(ANNOTATION_KEY, key);
+        table.getEAnnotations().add(result);
+        return result;
+    }
+
+    public static String getClassGroupsName(DAnnotation it) {
+        return it.getDetails().get(ANNOTATION_KEY);
+    }
+
+    public static List<DAnnotation> getClassGroupsByInitiale(Package root, DTable table) {
```
Comment (lredor) [inline plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/variables/VirtualTableService.java:108]:
```suggestion
    public static List<DAnnotation> getClassGroupsByInitial(Package root, DTable table) {
```

**Review by lredor — CHANGES_REQUESTED** [review]
Can you add this last line on all commits referencing issue [115] at the end of the commit message?
`Bug: https://github.com/eclipse-sirius/sirius-desktop/issues/115`

File: plugins/org.eclipse.sirius.doc/doc/Release_Notes.textile [hunk 33]
```diff
@@ -12,8 +12,23 @@ h3. User-Visible Changes
 
 h3. Specifier-Visible Changes
 
+* <span class="label label-info">Modified</span> In tables, expressions and tools are injected with more variables. 'table', 'containerView' and 'line' elements are now accessible. With those elements, Specifier can access to @AnnotationEntry@ elements of @DTable@ instances. This modification allows following use-cases: 
+** Create virtual lines: lines to group elements in tree with criteria based on attributes values or on user query.
+** Have columns unrelated to root element: for example, columns targeting ECore features to have dynamic feature table.
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.doc/doc/Release_Notes.textile:17]: Done

File: plugins/org.eclipse.sirius.doc/doc/Release_Notes.textile [hunk 34]
```diff
@@ -12,8 +12,23 @@ h3. User-Visible Changes
 
 h3. Specifier-Visible Changes
 
+* <span class="label label-info">Modified</span> In tables, expressions and tools are injected with more variables. 'table', 'containerView' and 'line' elements are now accessible. With those elements, Specifier can access to @AnnotationEntry@ elements of @DTable@ instances. This modification allows following use-cases:
```
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.doc/doc/Release_Notes.textile:15]: Done

File: plugins/org.eclipse.sirius.doc/doc/Release_Notes.textile (hunk 33 above)
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.doc/doc/Release_Notes.textile:17]: Feature Column are only used in EditionTable.
This was not possible for CrossTable.
I rewrote the sentence to be more precised.

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/variables/TableVariableTests.java
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/variables/TableVariableTests.java:1]: Done

File: plugins/org.eclipse.sirius.doc/doc/Release_Notes.textile
[the remark this answers is not in the document]
Comment (mypsycho) [reply plugins/org.eclipse.sirius.doc/doc/Release_Notes.textile:1]: Done

**Review by lredor — APPROVED** [review]

File: plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/variables/VirtualTableService.java (hunk 32 above)
Comment (mypsycho) [reply plugins/org.eclipse.sirius.tests.junit/src/org/eclipse/sirius/tests/unit/table/unit/variables/VirtualTableService.java:108]: Done.

**Comment by lredor** [comment]
PR to rebase manually before I can merge. I try to resolve "Release Notes" conflict on your branch by directly editing on web interface but without success.

**Comment by lredor** [comment]
This PR has been indirectly merged through PR #169

**Pull request closed without merging**